include/xtensor/xmath.hpp

```diff
@@ -3013,6 +3013,86 @@ namespace detail {
             return cov(eval(stack(xtuple(x, y))));
         }
     }
+
+    /*
+    * convolution mode placeholders
+    */
+    namespace conv_mode
+    {
+        struct valid{};
+        struct full {};
+    }
+
+    /*
+    * @brief computes the 1D convolution between two 1D expressions
+    *
+    * @param a 1D expression
+    * @param v 1D convolutional kernel
+    * @param mode placeholder from conv_mode namespace
+    * 
+    * @detail the algorithm convolves a with v and will incur a copy overhead
+    *   should v be longer than a.
+    */
+    template <class E1, class E2, class E3>
+    inline auto convolve(E1&& a, E2&& v, E3&& mode)
+    {
+        using value_type = typename std::decay<E1>::type::value_type;
+
+        if (a.shape().size() != 1 || v.shape().size() != 1)
+        {
+            XTENSOR_THROW(std::runtime_error, "Invalid dimentions convolution arguments must be 1D expressions");
+        }
+
+        if (a.shape(0) == 0 || v.shape(0) == 0)
+        {
+            XTENSOR_THROW(std::runtime_error, "Arguments cannot have zero length");
+        }
+
+        //swap them so a is always the longest one
+        if (a.shape(0) < v.shape(0))
+        {
+            auto temp = a;
+            a = v;
+            v = temp;
+        }
+
+        //this could benefit from C++17 if constexpr
```

Comment:
This can be achieved with tag dispatching (although I'm pretty sure the compiler can optimize the current implementation)

```cpp
template <class E1, class E2, class E3>
inline auto convolve_impl(E1&& e1, E2&& e2, E3 mode)
{
    XTENSOR_THROW(...);
}

template <class E1, class E2>
inline auto convolve_impl(E1&& e1, E2&& e2, conv_mode::full)
{
    // ...
}

template <class E1, class E2>
inline auto convolve_impl(E1&& e1, E2&& e2, conv_mode::valid)
{
    //....
}

Reply:
Done! Didn't implement the generic case as it would be better to have compile time error over runtime.